common/finalizer/finalizer.go

```diff
@@ -105,62 +106,65 @@ func (f *Finalizer) Run(
 	pool *goro.AdaptivePool,
 	timeout time.Duration,
 ) int {
-	if !f.finalized.CompareAndSwap(false, true) {
+	if timeout == 0 {
+		f.logger.Info("finalizer skipped: zero timeout")
+		return 0
+	}
+
+	f.mu.Lock()
+	if f.finalized {
 		f.logger.Warn("finalizer skipped: called more than once")
+		f.mu.Unlock()
 		return 0
 	}
+	f.finalized = true
+	f.mu.Unlock() // unlocking immediately to unblock any calls to Register/Deregister
 
-	if timeout == 0 {
-		f.logger.Warn("finalizer skipped: zero timeout")
+	totalCount := len(f.callbacks)
+	if totalCount == 0 {
+		f.logger.Debug("finalizer skipped: no callbacks")
 		return 0
 	}
 
-	f.mu.Lock()
-	defer f.mu.Unlock()
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
 
-	totalCount := len(f.callbacks)
 	f.logger.Info("finalizer starting",
 		tag.NewInt("items", totalCount),
 		tag.NewDurationTag("timeout", timeout))
 
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
-	defer cancel()
-
-	done := make(chan struct{})
 	completionChannel := make(chan struct{})
 	go func() {
+		defer func() { f.logger.Info("finalizer loop ended") }()
 		for _, callback := range f.callbacks {
 			// NOTE: Once `pool.Stop` is called due to a timeout, any remaining calls to `pool.Do` will do nothing.
 			pool.Do(func() {
 				defer func() { completionChannel <- struct{}{} }()
 				_ = callback(ctx)
 			})
 		}
-		done <- struct{}{}
 	}()
 
-	var completed int
-loop:
+	var completedCallbacks int
 	for {
 		select {
 		case <-completionChannel:
-			completed += 1
-			if completed == totalCount {
+			completedCallbacks += 1
+
+			if completedCallbacks == totalCount {
 				f.logger.Info("finalizer completed",
-					tag.NewInt("completed-items", completed))
-				break loop
+					tag.NewInt("completed", completedCallbacks))
+				return completedCallbacks
 			}
+
 		case <-ctx.Done():
 			pool.Stop()
+
 			f.logger.Error("finalizer timed out",
-				tag.NewInt("completed-items", completed),
-				tag.NewInt("unfinished-items", totalCount-completed))
-			break loop
+				tag.NewInt("completed", completedCallbacks),
+				tag.NewInt("unfinished", totalCount-completedCallbacks))
+
+			return completedCallbacks
 		}
 	}
-
-	// ensure that goroutine completed and does not access callbacks anymore before releasing the lock
-	<-done
```

Comment:
This is not needed anymore since Register/Deregister will not access or modify callbacks anymore once Run starts the goroutine.